### PR TITLE
INSP: Fix non_exhaustive false positive in same crate

### LINKED ISSUE #4896: False positive "Match must be exhaustive" when using #[non_exhaustive] in the same crate
## Environment

* **IntelliJ Rust plugin version:** 0.2.114.2151-193
* **Rust toolchain version:** 1.40.0
* **IDE name and version:** CLion 2019.3.3
* **Operating system:** Linux (Fedora 31)

## Problem description
`#[non_exhaustive]` was introduced in [Rust 1.40.0](https://blog.rust-lang.org/2019/12/19/Rust-1.40.0.html#[non_exhaustive]-structs,-enums,-and-variants). It prevents users of a crate from exhaustively matching on whatever is annotated with `#[non_exhaustive]`. However, that restriction does not apply to the crate in which the type is defined. [Link to section in RFC which mentions this](https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md#enums-1)

IntelliJ Rust correctly identifies that matching on a `#[non_exhaustive]` type is prohibited, but it continues to throw that error when matching on the type inside the declaring crate.

## Steps to reproduce

```rust
#[non_exhaustive]
enum Error {
    Variant
}

fn main() {
    let error = Error::Variant;

    match error {
        Error::Variant => println!("Variant")
    }
}
```

image

### PR DESCRIPTION
**Pull request by Mcat12**
Fixes #4896 by checking if the enum was declared in the same crate or not. I'm not sure if using `crateRoot` is the right approach, so feedback is welcome!

### CODE REVIEW

File: src/test/kotlin/org/rust/ide/inspections/RsMatchCheckInspectionTest.kt [hunk 1]
```diff
@@ -1227,4 +1227,39 @@ class RsMatchCheckInspectionTest : RsInspectionsTestBase(RsMatchCheckInspection:
...
+        fn main() {
+            let error = Error::Variant;
+
+            match error {
+                Error::Variant => println!("Variant")
+            }
+        }
+    """)
+
+    fun `test non-exhaustive enum match in different crate`() = checkByFileTree("""
```
Comment (Mcat12) [inline src/test/kotlin/org/rust/ide/inspections/RsMatchCheckInspectionTest.kt:1246]: This test currently fails. It doesn't run the inspection code, so I probably don't have it set up correctly.

Comment (Stzx) [reply src/test/kotlin/org/rust/ide/inspections/RsMatchCheckInspectionTest.kt:1246]: Hey, let's replace it with the new code below:

```kotlin
    @ProjectDescriptor(WithDependencyRustProjectDescriptor::class)
    fun `test non-exhaustive enum match in different crate`() = checkByFileTree("""
    //- main.rs
        extern crate dep_lib_target;
        use dep_lib_target::Error;

        fn main() {
            let error = Error::Variant;

            <error descr="Match must be exhaustive [E0004]">match/*caret*/</error> error {
                Error::Variant => println!("Variant")
            }
        }
    //- dep-lib/lib.rs
        #[non_exhaustive]
        pub enum Error {
            Variant
        }
    """)
```

**Comment by Mcat12** [comment]
> org.rust.toml.resolve.CargoTomlKeyResolveTest > test in dependencies block FAILED
>     com.intellij.testFramework.LoggedErrorProcessor$TestLoggerAssertionError at LoggedErrorProcessor.java:56
>         Caused by: java.lang.Throwable at Logger.java:146

https://github.com/intellij-rust/intellij-rust/pull/5466/checks?check_run_id=721287533
Is this error spurious?

**Comment by Undin** [comment]
@Mcat12 yep

**Comment by Mcat12** [comment]
Restarting CI to see if that spurious error will go away.

**Comment by Mcat12** [comment]
A different, unrelated error:
```
org.rust.lang.core.macros.RsMacroExpansionCachingTest > test edit def 1 FAILED
    java.lang.IllegalStateException: Expected to re-expand [foo], re-expanded [foo, bar] instead
        at org.rust.lang.core.macros.RsMacroExpansionCachingTest.checkReExpandedInner(RsMacroExpansionCachingTest.kt:78)
        at org.rust.lang.core.macros.RsMacroExpansionCachingTest.checkReExpanded(RsMacroExpansionCachingTest.kt:87)
        at org.rust.lang.core.macros.RsMacroExpansionCachingTest.test edit def 1(RsMacroExpansionCachingTest.kt:134)
```

**Comment by ortem** [comment]
@Mcat12 Thank you and welcome to IJ Rust contributors! 🎉 

bors r+

**Pull request merged**